bitcoinops/main.py

```diff
@@ -0,0 +1,145 @@
+                continue
+            for post_file in files:
+                logger.info("Parsing {}".format(join(file[0], post_file)))
+                documents.append(parse_post(join(file[0], post_file), typeof))
+            return documents
+
+    else:
+        logger.critical("Data Directory not available.")
+
+
+def parse_post(post_file: str, typeof: str):
+    try:
+        with open(post_file, 'r', encoding='utf-8') as file:
+            content = file.read()
+        content = re.sub(r'{%.*%}', '', content, flags=re.MULTILINE)
+        front_matter, body = parse_markdowns(content)
+        metadata = yaml.safe_load(front_matter)
+        custom_id = basename(post_file).replace('.md', '') if typeof == 'topic' else metadata['slug']
+        document = {
+            "id": "bitcoinops-" + custom_id,
+            "title": metadata['title'],
+            "body_formatted": body,
+            "body": body,
+            "body_type": "markdown",
+            "created_at": metadata['date'].strftime('%Y-%m-%dT%H:%M:%S.000Z') if metadata.get('date') else None,
+            "domain": "https://bitcoinops.org/en/",
+            "url": "https://bitcoinops.org/en/topics/{}".format(
+                custom_id + '.md') if typeof == "topic" else "https://bitcoinops.org${}".format(metadata['permalink']),
+            "type": "topic" if typeof == "topic" else metadata['type'],
+            "language": metadata['lang'] if 'lang' in metadata.keys() else 'en',
+            "authors": ["bitcoinops"],
+            "indexed_at": datetime.now().isoformat()
+        }
+        return document
+    except IOError as e:
+        logger.warning("Issue while parsing the file, {}".format(post_file))
+
+
+def parse_markdowns(content: str):
+    lines = content.split("\n")
+    front_matter = False
+    body_flag = False
+    front_matter_list = []
+    body_list = []
+
+    for line in lines:
+        if line.startswith("---"):
+            if front_matter:
+                front_matter = False
+                body_flag = True
+                continue
+            if body_flag:
+                break
+            front_matter = True
+            continue
+        if front_matter:
+            front_matter_list.append(line)
+        elif body_flag:
+            body_list.append(line)
+    return '\n'.join(front_matter_list), '\n'.join(body_list)
+
+
+def parseTopics():
+    pass
+
+
+def main(data_source_url: str, download_dir: str, file_name: str, postdir_path: str, topicdir_path: str) -> None:
+    if download_zip(data_source_url, download_dir, file_name):
+        all_post = dir_walk(join(download_dir, file_name[:-4:], postdir_path), "posts")
+        all_topic = dir_walk(join(download_dir, file_name[:-4:], topicdir_path), "topic")
+        all_post.extend(all_topic)
+        cnt = 0
+        for post in all_post:
+            try:
+                res = upsert_document(index_name=getenv('INDEX'), doc_id=post['id'], doc_body=post)
```

Comment:
@urvishp80 do I understand correctly here that the previous scraper for Optech was not updating the pages, and this one does? Which is related with https://github.com/bitcoinsearch/scraper/issues/68

Reply:
Yes @kouloumos , `upsert` function update the passed fields and keep the remaining ones as it is in an Elasticsearch Index.

Reply:
But @urvishp80 is it true that our current logic doesn't update the Optech pages?

Reply:
@kouloumos The current logic uses `index` function which means, it will add a new document or replaces an existing document with the same ID. On the other hand, `upsert` function updates a document if it exists for the given ID; otherwise, inserts a new  one.